### ISSUE TITLE
No metric data when running in remote mode

### ISSUE THREAD

**Issue by drasmuss** [microsoft/nni#2408]
**Short summary about the issue/question**:

When running an experiment in remote mode the remote jobs complete normally, and I can see them reporting their results (on the remote side in the `trialkeeper_stdout` file):
```
...
NNISDK_MEb'{"parameter_id": 0, "trial_job_id": "nV851", "type": "FINAL", "sequence": 0, "value": "1.4060341119766235"}'
[2020-05-06 17:36:59.714002] INFO subprocess terminated. Exit code is 0. Quit
[2020-05-06 17:37:09.722537] INFO NNI trial keeper exit with code 0
```
But no metric data shows up locally:
<img width="1150" alt="Capture" src="https://user-images.githubusercontent.com/1952220/81211516-df7d5780-8fa9-11ea-8993-3803f523efcb.PNG">
And if I check http://127.0.0.1:8080/api/v1/nni/metric-data the result is also empty `[]`.

**Brief what process you are following**: 

Executing `nnictl create --config nni-config/config.yml --debug`

with `config.yml`
```
...
trainingServicePlatform: remote
searchSpacePath: search_space.json
...
tuner:
  builtinTunerName: SMAC
  includeIntermediateResults: true
  classArgs:
    optimize_mode: minimize
assessor:
  builtinAssessorName: Curvefitting
  classArgs:
    optimize_mode: minimize
    epoch_num: 10
    threshold: 1.1
    start_step: 5
trial:
  command: python myfile.py
  codeDir: ../code_dir
machineList:
  - ip: ...
    username: ...
    sshKeyPath: ...
    useActiveGpu: false
```

Connecting from my local machine to remote Azure VMs.

**How to reproduce it**: 

**nni Environment**:
- nni version: 1.5
- nni mode(local|pai|remote): remote
- OS: Ubuntu 16.04 running in Windows 10 WSL (local) and Ubuntu 18.04 (remote)
- python version: 3.7
- is conda or virtualenv used?: conda
- is running in docker?: no

**need to update document(yes/no)**: no

**Anything else we need to know**:

**Log message**:
 - [nnimanager.log and dispatcher.log](https://github.com/microsoft/nni/blob/master/docs/en_US/Tutorial/HowToDebug.md#experiment-root-directory) : 

[nnimanager.log](https://github.com/microsoft/nni/files/4588638/nnimanager.log)

[dispatcher.log](https://github.com/microsoft/nni/files/4588640/dispatcher.log)

 - [nnictl stdout and stderr](https://github.com/microsoft/nni/blob/master/docs/en_US/Tutorial/Nnictl.md#nnictl%20log%20stdout) : 

stdout
```
-----------------------------------------------------------------------
                Experiment start time 2020-05-06 14:31:42
-----------------------------------------------------------------------
```

stderr
```
-----------------------------------------------------------------------
                Experiment start time 2020-05-06 14:31:42
-----------------------------------------------------------------------
(node:2447) ExperimentalWarning: The fs.promises API is experimental
```

In the `trialkeeper_stderr` file of the remote trials I see this, not sure if it is related:
```
HTTPConnectionPool(host='192.168.0.184', port=8081): Max retries exceeded with url: /api/v1/nni-pai/stdout/M5t8IBz3/nV851 (Caused by ConnectTimeoutError(<urllib3.connection.HTTPConnection object at 0x7f6084ab4dd8>, 'Connection to 192.168.0.184 timed out. (connect timeout=10)'))
HTTPConnectionPool(host='192.168.0.184', port=8081): Max retries exceeded with url: /api/v1/nni-pai/stdout/M5t8IBz3/nV851 (Caused by ConnectTimeoutError(<urllib3.connection.HTTPConnection object at 0x7f6084b187b8>, 'Connection to 192.168.0.184 timed out. (connect timeout=10)'))
HTTPConnectionPool(host='192.168.0.184', port=8081): Max retries exceeded with url: /api/v1/nni-pai/stdout/M5t8IBz3/nV851 (Caused by ConnectTimeoutError(<urllib3.connection.HTTPConnection object at 0x7f6084ad22b0>, 'Connection to 192.168.0.184 timed out. (connect timeout=10)'))
HTTPConnectionPool(host='192.168.0.184', port=8081): Max retries exceeded with url: /api/v1/nni-pai/stdout/M5t8IBz3/nV851 (Caused by ConnectTimeoutError(<urllib3.connection.HTTPConnection object at 0x7f6084ad29b0>, 'Connection to 192.168.0.184 timed out. (connect timeout=10)'))
HTTPConnectionPool(host='192.168.0.184', port=8081): Max retries exceeded with url: /api/v1/nni-pai/stdout/M5t8IBz3/nV851 (Caused by ConnectTimeoutError(<urllib3.connection.HTTPConnection object at 0x7f6084adf128>, 'Connection to 192.168.0.184 timed out. (connect timeout=10)'))
HTTPConnectionPool(host='192.168.0.184', port=8081): Max retries exceeded with url: /api/v1/nni-pai/stdout/M5t8IBz3/nV851 (Caused by ConnectTimeoutError(<urllib3.connection.HTTPConnection object at 0x7f6084adf860>, 'Connection to 192.168.0.184 timed out. (connect timeout=10)'))
```

**Comment by SubodhDahal** [comment]
I have encountered the same issue. I have a similar config, but I am using **virtualenv** and the remote machine is running on Arch Linux. I am on Pop!-OS 18.04.

I also had to specify the `nniManagerIp` because I don't have an `eth0` interface locally. However, specifying the value as anything other than the remote machine IP didn't even let the dispatcher to be executed.

**Comment by seanytak** [comment]
Hello, I am also experiencing this issue albeit using the TrainingService option for FrameworkController with Azure Kubernetes Service. I am working off of an Azure Data Science Virtual Machine as my local machine. Explicitly setting my local machine's ip via `nniManagerIP` does not seem to fix my issue.

**Comment by SparkSnail** [comment]
Hi all, this issue seems familiar in non-local platforms. In NNI's experiment of non-local platform, NNI will start two restful server. One restful server is used to server WebUI and NNICTL, the port of this restful server is configured in nnictl command using `nnictl --port`, and the default value is 8080. Another restful server is used to accept metrics data from remote machines, and show the metrics data in WebUI, the port of this restful server is `{port}+1`, this `{port}` is the `nnictl --port` just mentioned, and does not support user-defined. As you meet connection error in `trialkeeper_stderr`, there are two ways for fixing:  
1. Check if your nniManagerIp is correct.
2. Check if your `{port}+1` is accessible, and not blocked by firewall.

**Comment by seanytak** [comment]
Fixing the inbound port rules worked for my Azure VM. Is there a way for NNI to note if the port is blocked vs. the host is unreachable (bad IP). I guess this is more of a general networking question but since I was unaware of the `{port}+1` needing to be open as well I didn't try that for debugging. Thanks!

**Comment by SparkSnail** [comment]
Hi @seanytak, I've add this issue to NNI's FAQ page. https://github.com/microsoft/nni/wiki/FAQ#no-metric-data-when-running-in-non-local-mode

**Comment by drasmuss** [comment]
Fixing the port/IP issues solved this for me as well. I ended up solving it by running the NNI manager in a VM within the same vnet as all the other VMs (and then I connect to that VM from my local machine), which worked better than having all the VMs trying to communicate with my local machine. Thanks!

**Comment by mbu93** [comment]
The aforementioned solutions don't seem to work for me. I am trying to execute the mnist-pytorch example with the following setup:
NNI (Version 1.9) on remote mode with the following config:

```yaml
authorName: default
experimentName: example_mnist_pytorch
trialConcurrency: 1
maxExecDuration: 1h
nniManagerIp: 0.0.0.0
maxTrialNum: 10
trainingServicePlatform: remote
searchSpacePath: search_space.json
useAnnotation: false
tuner:
  builtinTunerName: TPE
  classArgs:
    optimize_mode: maximize
trial:
  command: sudo sh -c "python3 mnist.py > /proc/1/fd/1"
  codeDir: .
  gpuNum: 0
machineList:
  - ip: 172.17.0.1
    username: ssh********
    passwd: ******
    port: 22
```
where my "remote" machine is a docker container running an openssh-server that is currently spawned locally for the sake of 
rapid testing. The container is based on the latest msranni:nnimanager image. I tried both forwarding only the required ports, as well as the whole host's network as in the current config. 
Running 
```
nnictl create -c config.yml --debug
```
will lead me to some default stuff like 
```shell
INFO:  expand searchSpacePath: search_space.json to /home/mbu93/Projects/openvaccine-covid19/nni/examples/trials/mnist-pytorch/search_space.json
INFO:  expand codeDir: . to /home/mbu93/Projects/openvaccine-covid19/nni/examples/trials/mnist-pytorch/.
INFO:  Starting restful server...
INFO:  Successfully started Restful server!
INFO:  Setting remote config...
INFO:  Successfully set remote config!
INFO:  Starting experiment...
INFO:  Successfully started experiment!
------------------------------------------------------------------------------------
The experiment id is AAwJ3qgC
The Web UI urls are: 0.0.0.0:8080
------------------------------------------------------------------------------------

You can use these commands to get more information about the experiment
------------------------------------------------------------------------------------
         commands                       description
1. nnictl experiment show        show the information of experiments
2. nnictl trial ls               list all of trial jobs
3. nnictl top                    monitor the status of running experiments
4. nnictl log stderr             show stderr log content
5. nnictl log stdout             show stdout log content
6. nnictl stop                   stop an experiment
7. nnictl trial kill             kill a trial job by id
8. nnictl --help                 get help information about nnictl
------------------------------------------------------------------------------------
Command reference document https://nni.readthedocs.io/en/latest/Tutorial/Nnictl.html
------------------------------------------------------------------------------------
```
Additionally, checking the logs of the container shows the trial is in fact running (I've skipped the training code and simply report a final result of 0.0..doesn't work in neither of the cases anyway):
```shell
[11/23/2020, 02:32:39 PM] WARNING (nni) Requesting parameter without NNI framework, returning empty dict
{'data_dir': './data', 'batch_size': 64, 'batch_num': None, 'hidden_size': 512, 'lr': 0.01, 'momentum': 0.5, 'epochs': 10, 'seed': 1, 'no_cuda': False, 'log_interval': 1000}
[11/23/2020, 02:32:39 PM] INFO (nni) Final result: 0.0
[11/23/2020, 02:33:00 PM] WARNING (nni) Requesting parameter without NNI framework, returning empty dict
{'data_dir': './data', 'batch_size': 64, 'batch_num': None, 'hidden_size': 512, 'lr': 0.01, 'momentum': 0.5, 'epochs': 10, 'seed': 1, 'no_cuda': False, 'log_interval': 1000}
[11/23/2020, 02:33:00 PM] INFO (nni) Final result: 0.0
```
Checking the trialkeeper stderr and stdout confirms everything seems to be working fine:
/tmp/nni-experiments/AAwJ3qgC/trials/BB4yX/trialkeeper_stdout
```
[2020-11-23 14:31:08.670579] WARNING Skipping version check!
[2020-11-23 14:31:08.672473] INFO Trial keeper spawns a subprocess (pid 5304) to run command: ['CUDA_VISIBLE_DEVICES= ', 'sudo', 'sh', '-c', 'python3 mnist.py > /proc/1/fd/1']
[2020-11-23 14:31:20.687251] INFO subprocess terminated. Exit code is 0. Quit
[2020-11-23 14:31:20.700222] INFO NNI trial keeper exit with code 0
```
/tmp/nni-experiments/AAwJ3qgC/trials/BB4yX/trialkeeper_stderr
```
```
I also double-checked, that choosing a non-resolvable address would cause an entry in /tmp/nni-experiments/AAwJ3qgC/trials/BB4yX/trialkeeper_stderr. Both of the required rest-server ports (8080 and 8081) are open and can be accessed e.g. using curl.
A thing I realized is, that the metrics file seems to be empty for all trials. The trials still get marked as succeeded in the webui (running on the host). 

After debugging this for around half a day I have no idea what's wrong here. I also don't really understand how metrics will be reported. If I've got this right we are talking about a simple API post like xxxxxx/api/v1/nni/update-metrics, where the actual value of the trial metric will be part of the JSON body? Could you provide some example for such an API call, so that I can manually test updating the trial metrics. That would be great. 

Thanks for your affords!
KR
